tokio/src/runtime/mod.rs

```diff
@@ -280,23 +281,23 @@ pub struct Runtime {
     handle: Handle,
 
     /// Blocking pool handle, used to signal shutdown
-    blocking_pool: BlockingPool,
+    blocking_pool: Arc<BlockingPool>,
 }
 
 /// The runtime executor is either a thread-pool or a current-thread executor.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 enum Kind {
     /// Not able to execute concurrent tasks. This variant is mostly used to get
     /// access to the driver handles.
-    Shell(Shell),
+    Shell(Arc<Mutex<Option<Shell>>>),
 
     /// Execute all tasks on the current-thread.
     #[cfg(feature = "rt-core")]
-    Basic(BasicScheduler<time::Driver>),
+    Basic(Arc<Mutex<Option<BasicScheduler<time::Driver>>>>),
 
     /// Execute tasks across multiple threads.
     #[cfg(feature = "rt-threaded")]
-    ThreadPool(ThreadPool),
+    ThreadPool(Arc<ThreadPool>),
```

Comment:
Not really a blocker, since we can always fix this later, but it seems like we can fold these together into one `Arc`, right? 

```rust
pub struct Runtime {
    inner: Arc<Inner>,
    // ...
}

struct Inner {
    blocking_pool: BlockingPool,
    kind: Kind,
}

enum Kind {
    Shell(Mutex<Option<Shell>>),
    Basic(Mutex<Option<BasicScheduler<time::Driver>>>),
    ThreadPool(ThreadPool),
}
```

That way we'd only bump one `Arc` when cloning this thing. 

AFAICT, there aren't any places where the scheduler or blocking pool could outlive the other but i may be overlooking it...

Reply:
Yes, next PR is to refactor the `BasicScheduler` and `Shell` so we can just have one arc. Avoiding that for now to reduce scope of this PR.

Reply:
sgtm

Reply:
I commented above, but I don't think we need the `Arc` at all. `Runtime` doesn't need to be `Clone`, just be `Sync`.

Reply:
@carllerche is this to let the user arc the runtime?

Reply:
Yes. The user can opt into the single `Arc` if they  need to.